71_RayTracingPipeline/app_resources/raytrace.rchit.hlsl

```diff
@@ -0,0 +1,157 @@
+    switch (objType)
+    {
+        case OT_CUBE:
+        {
+                uint32_t v0 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i0 * vertexStride, 2u);
+                uint32_t v1 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i1 * vertexStride, 2u);
+                uint32_t v2 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i2 * vertexStride, 2u);
+
+                n0 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v0).xyz);
+                n1 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v1).xyz);
+                n2 = normalize(nbl::hlsl::spirv::unpackSnorm4x8(v2).xyz);
+            }
+            break;
+        case OT_SPHERE:
+        case OT_CYLINDER:
+        case OT_ARROW:
+        case OT_CONE:
+        {
+                uint32_t v0 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i0 * vertexStride);
+                uint32_t v1 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i1 * vertexStride);
+                uint32_t v2 = vk::RawBufferLoad < uint32_t > (normalVertexBufferAddress + i2 * vertexStride);
+
+                n0 = normalize(unpackNormals3x10(v0));
+                n1 = normalize(unpackNormals3x10(v1));
+                n2 = normalize(unpackNormals3x10(v2));
+            }
+            break;
+        case OT_RECTANGLE:
+        case OT_DISK:
+        case OT_ICOSPHERE:
+        default:
+        {
+                n0 = normalize(vk::RawBufferLoad <
+                float3 > (normalVertexBufferAddress + i0 * vertexStride));
+                n1 = normalize(vk::RawBufferLoad <
+                float3 > (normalVertexBufferAddress + i1 * vertexStride));
+                n2 = normalize(vk::RawBufferLoad <
+                float3 > (normalVertexBufferAddress + i2 * vertexStride));
+            }
+    }
+
+    float3 barycentrics = float3(0.0, bary);
+    barycentrics.x = 1.0 - barycentrics.y - barycentrics.z;
+
+    VertexData data;
+    data.position = barycentrics.x * p0 + barycentrics.y * p1 + barycentrics.z * p2;
+    data.normal = normalize(barycentrics.x * n0 + barycentrics.y * n1 + barycentrics.z * n2);
+    return data;
+}
+
+[shader("closesthit")]
+void main(inout ColorPayload p, in BuiltInTriangleIntersectionAttributes attribs)
+{
+    const int instID = InstanceID();
+    const int primID = PrimitiveIndex();
+    const STriangleGeomInfo geom = vk::RawBufferLoad < STriangleGeomInfo > (pc.triangleGeomInfoBuffer + instID * sizeof(STriangleGeomInfo));
+    const VertexData vertexData = fetchVertexData(instID, primID, geom, attribs.barycentrics);
+    const float32_t3 worldPosition = mul(ObjectToWorld3x4(), float32_t4(vertexData.position, 1));
+    const float32_t3 worldNormal = normalize(mul(vertexData.normal, WorldToObject3x4()).xyz);
+
+    RayLight cLight;
+    cLight.inHitPosition = worldPosition;
+    CallShader(pc.light.type, cLight);
+
+    float32_t3 diffuse = computeDiffuse(geom.material, cLight.outLightDir, worldNormal);
```

Comment:
`callShader` is a invocation reorder instruction (function call) it can suspend your current subgroup invocation, and then resume the epilogue (the bit that comes after the function call) on another subgroup.

This by its nature requires the spilling of all live variables (variables being stored to before the call and read after) to a stack.

Its best practice to keep them to a minimum and only use `CallShader` as a tail call if possible

Reply:
half of the stuff in your payload could be loaded from memory/computed after the shadow ray is known 

to decide whether to cast a shadow ray and cast it, just the hit position and normal are needed

Reply:
done by me